### DISCUSSION TITLE
How to use DetermineBondOrders()

### DISCUSSION THREAD

**Discussion by peastman** [rdkit/rdkit#7289] [Q&A]
I'm trying to use `DetermineBondOrders()`, but I can't figure out how to make it work correctly.  The following script builds a methane molecule and tries to determine the bond orders.

```python
from rdkit import Chem
from rdkit.Chem import rdDetermineBonds

elements = [6, 1, 1, 1, 1]
bonds = [(0, 1), (0, 2), (0, 3), (0,4)]

mol = Chem.EditableMol(Chem.Mol())
for e in elements:
    a = Chem.Atom(e)
    a.SetNoImplicit(True)
    mol.AddAtom(a)
for bond in bonds:
    mol.AddBond(bond[0], bond[1])
mol = mol.GetMol()
rdDetermineBonds.DetermineBondOrders(mol, charge=0, embedChiral=False)
for bond in mol.GetBonds():
    print(bond.GetBondType())
```

It fails to determine them.  Here is the output.

```
UNSPECIFIED
UNSPECIFIED
UNSPECIFIED
UNSPECIFIED
```

Can anyone tell me what I'm doing wrong?

Thanks!

**Comment by greglandrum** [comment]
Hi @peastman, that's an interesting use case that I don't think we necessarily planned for. It is an interesting one though and it would be nice if it actually worked.

The use case we envisioned is that you start with a molecule that has 3D coordinates, and then call either `rdDetermineBonds.DetermineConnectivity()` - which sets all the bonds it finds to single - followed by `rdDetermineBonds.DetermineBondOrders()` - which attempts to set correct bond orders - OR just do it all in a single step by calling `rdDetermineBonds.DetermineBonds()`

I will fix the bug you reported w.r.t. the handling of unspecified bonds (#7299), but in the meantime, I think you can probably make your code work by always setting bonds to SINGLE (you do this with the optional argument to `AddBond()`). That will provide `DetermineBondOrders()` the input it expects. Obviously in the case of methane you won't see anything, but it does work for the example from #7299:
```
In [44]: from rdkit import Chem
    ...: from rdkit.Chem import rdDetermineBonds
    ...: 

In [45]: elements = [6, 8, 8, 8, 1, 1]
    ...: bonds = [(0, 1), (0, 2), (0, 3), (2, 4), (3, 5)]
    ...: 

In [46]: mol = Chem.EditableMol(Chem.Mol())
    ...: for e in elements:
    ...:     mol.AddAtom(Chem.Atom(e))
    ...: for bond in bonds:
    ...:     mol.AddBond(bond[0], bond[1], Chem.BondType.SINGLE)
    ...: mol = mol.GetMol()
    ...: rdDetermineBonds.DetermineBondOrders(mol, charge=0, embedChiral=False)
    ...: 
    ...: for bond in mol.GetBonds():
    ...:     print(bond.GetBondType())
    ...: 
DOUBLE
SINGLE
SINGLE
SINGLE
SINGLE
```

**Comment by peastman** [comment]
Thanks!  I'd also tried the pathway you suggest of loading coordinates and letting it identify the bonds automatically.  The problem I had with that is that it isn't reliable.  It sometimes misidentifies which atoms are bonded to each other.  Since I already know exactly which ones are bonded, I hoped this would be more robust by eliminating a potential point of failure.

By the way, this is also related to #6501, which is another manifestation of the same problem: the bonds are already known, but it ignores the known bonds, tries to identify them based on coordinates, and sometimes gets them wrong.

**Comment by peastman** [comment]
I wrote the following routine to test it out.  It takes a SMILES string and creates a Mol from it.  It then creates a second Mol by copying over just the atoms and bonds and calls `DetermineBondOrders()` to fill in the missing information, and checks whether the two molecules match.

```python
def compare_molecules(smiles):
    original_mol = Chem.MolFromSmiles(smiles)
    original_mol = Chem.AddHs(original_mol)
    new_mol = Chem.EditableMol(Chem.Mol())
    for atom in original_mol.GetAtoms():
        a = Chem.Atom(atom.GetAtomicNum())
        a.SetNoImplicit(True)
        new_mol.AddAtom(a)
    for bond in original_mol.GetBonds():
        new_mol.AddBond(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), Chem.BondType.SINGLE)
    new_mol = new_mol.GetMol()
    charge = sum(atom.GetFormalCharge() for atom in original_mol.GetAtoms())
    rdDetermineBonds.DetermineBondOrders(new_mol, charge=charge, embedChiral=False)
    original_smiles = Chem.MolToSmiles(original_mol)
    new_smiles = Chem.MolToSmiles(new_mol)
    if original_smiles != new_smiles:
        print('Error:', smiles)
        print(original_smiles)
        print(new_smiles)
        return False
    return True
```

I tested it on thousands of SMILES strings drawn from PubChem, and over 90% of them give a perfect match, so that's encouraging.  Even when they don't match, it isn't clear to me that the result is wrong.  I'm not a chemist, but in a lot of cases the result of `DetermineBondOrders()` looks more plausible to me than the original description from PubChem.

There are some clear problems though.  I found a couple of cases where it just hangs.  It seems to be going into an infinite loop or something.  Try this:

```python
compare_molecules('[N-]=[N+]=NCC1OC(OS(=O)(=O)[O-])C(OS(=O)(=O)[O-])C(OS(=O)(=O)[O-])C1OS(=O)(=O)[O-]')
```

It hangs and I have to force quit the process.

There also are some cases where the results look very questionable to me.  For example, try this:

```python
compare_molecules('[NH-]c1n[n+](-c2ccc(Cl)c(Cl)c2)no1')
```

It produces the output

```
Error: [NH-]c1n[n+](-c2ccc(Cl)c(Cl)c2)no1
[H][N-]c1n[n+](-c2c([H])c([H])c(Cl)c(Cl)c2[H])no1
[H][N-]C1=NN(C2=C([H])C(Cl)=C(Cl)C([H])=C2[H])N=[O+]1
False
```

It puts the positive charge onto the oxygen rather than the nitrogen, which seems unlikely to be correct.

**Comment by peastman** [comment]
I'm also finding a lot of cases where it fails because it makes invalid assumptions about the possible valence of different elements.

```
>>> compare_molecules('[O-][I+3]([O-])([O-])[O-]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 13, in compare_molecules
ValueError: Valence of atom 1 is 4, which is larger than the allowed maximum, 1
>>> compare_molecules('O=NO[Cl+][O-]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 13, in compare_molecules
ValueError: Valence of atom 3 is 2, which is larger than the allowed maximum, 1
>>> compare_molecules('F[P-](F)(F)(F)(F)F')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 13, in compare_molecules
ValueError: Valence of atom 1 is 6, which is larger than the allowed maximum, 3
```
